File: src/main/ruby/g0001_0100/s0084_largest_rectangle_in_histogram/solution.rb
# #Hard #Top_100_Liked_Questions #Top_Interview_Questions #Array #Stack #Monotonic_Stack
# #Big_O_Time_O(n_log_n)_Space_O(log_n) #2023_11_22_Time_197_ms_(98.25%)_Space_223.4_MB_(24.56%)

# @param {Integer[]} heights
# @return {Integer}
def largest_rectangle_area(heights)
  area = 0
  stack = []
  return 0 if heights.empty?
  return heights[0] * heights.length if heights.uniq.length == 1

  heights.each.with_index do |height, index|
    start = index
    while !stack.empty? && stack.last[1] > height do
      i, h = stack.pop
      area = [area, h * (index - i)].max
      start = i
    end
    stack << [start, height]
  end

  stack.each do |i, s|
    area = [area, s * (heights.length - i)].max
  end

  area
end

File: src/main/js/g0001_0100/s0084_largest_rectangle_in_histogram/solution.js
// #Hard #Top_100_Liked_Questions #Top_Interview_Questions #Array #Stack #Monotonic_Stack
// #Big_O_Time_O(n_log_n)_Space_O(log_n) #2024_12_12_Time_11_ms_(96.58%)_Space_61.2_MB_(84.95%)

/**
 * @param {number[]} heights
 * @return {number}
 */
var largestRectangleArea = function(heights) {
    let stack = []
    let maxArea = 0
    let index = 0

    while (index < heights.length) {
        if (stack.length === 0 || heights[stack[stack.length - 1]] <= heights[index]) {
            stack.push(index++)
        } else {
            let top = stack.pop()
            let area = heights[top] * (stack.length === 0 ? index : index - stack[stack.length - 1] - 1)
            maxArea = Math.max(maxArea, area)
        }
    }

    while (stack.length > 0) {
        let top = stack.pop()
        let area = heights[top] * (stack.length === 0 ? index : index - stack[stack.length - 1] - 1)
        maxArea = Math.max(maxArea, area)
    }

    return maxArea
};

export { largestRectangleArea }
